internal/queue/consumer.go

```diff
@@ -0,0 +1,286 @@
+type JobDatabase interface {
+	AddJob(ctx context.Context, job *database.Job) error
+	UpdateJobStarted(ctx context.Context, platform, id string, startedAt time.Time, raw map[string]any) error
+	UpdateJobCompleted(ctx context.Context, platform, id string, completedAt time.Time, raw map[string]any) error
+}
+
+// AmqpConsumer is an AMQP consumer for workflow job events.
+type AmqpConsumer struct {
+	amqpChannel    amqpChannel
+	amqpConnection amqpConnection
+	connectFunc    func(uri string) (amqpConnection, error)
+	uri            string
+	queueName      string
+	db             JobDatabase
+	mu             sync.Mutex
+	logger         *slog.Logger
+}
+
+// NewAmqpConsumer creates a new AmqpConsumer with the given dependencies.
+func NewAmqpConsumer(uri, queueName string, db JobDatabase, logger *slog.Logger) *AmqpConsumer {
+	if logger == nil {
+		logger = slog.Default()
+	}
+	return &AmqpConsumer{
+		uri:         uri,
+		queueName:   queueName,
+		connectFunc: amqpConnect,
+		db:          db,
+		logger:      logger,
+	}
+}
+
+// Start begins consuming messages from the AMQP queue and processing them.
+func (c *AmqpConsumer) Start(ctx context.Context) error {
+	err := c.resetConnectionOrChannelIfNecessary()
+	if err != nil {
+		return err
+	}
+
+	msgs, err := c.consumeMsgs()
+	if err != nil {
+		return err
+	}
+
+	c.logger.Info("AMQP consumer started", "queue", c.queueName)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg, ok := <-msgs:
+			if !ok {
+				return fmt.Errorf("messages channel closed")
+			}
+			if err := c.handleMessage(ctx, msg); err != nil {
+				c.logger.Error("failed to handle message", "error", err)
+			}
+		}
+	}
+}
+
+// WorkflowJobWebhook represents the structure of a workflow job webhook message.
+type WorkflowJobWebhook struct {
+	Action      string      `json:"action"`
+	WorkflowJob WorkflowJob `json:"workflow_job"`
+}
+
+// WorkflowJob represents the structure of a workflow job in the webhook message.
+type WorkflowJob struct {
+	ID          int      `json:"id"`
+	Labels      []string `json:"labels"`
+	Status      string   `json:"status"`
+	CreatedAt   *string  `json:"created_at"`
+	StartedAt   *string  `json:"started_at,omitempty"`
+	CompletedAt *string  `json:"completed_at,omitempty"`
+}
+
+// handleMessage processes a single AMQP message.
+func (c *AmqpConsumer) handleMessage(ctx context.Context, msg amqp.Delivery) error {
+	webhook := WorkflowJobWebhook{}
+	if err := json.Unmarshal(msg.Body, &webhook); err != nil {
+		msg.Nack(false, false) // don't requeue
+		return fmt.Errorf("failed to unmarshal message body: %w", err)
+	}
+	switch webhook.Action {
+	case "queued": // other possible action: "waiting"
+		return c.insertJobToDB(ctx, webhook.WorkflowJob, msg)
+	case "in_progress":
+		return c.updateJobStartedInDB(ctx, webhook.WorkflowJob, msg)
+	case "completed":
+		return c.updateJobCompletedInDB(ctx, webhook.WorkflowJob, msg)
```

Comment:
I don't really like passing the object (to ack and nack) so deep in the business logic. I'd rather manage every ack/nack from the handle message side (easier for maintenance and readability).
So I would use the potential error returned by the `insert` and `update` methods to know if I ack or nack and I requeue

Reply:
In the same spirit, I'd avoid passing the `amqp.Delivery` to the database methods, but pass the message body/headers to these methods. Will be easier to unit test, will even allow to decouple the DB operations from AmqpConsumer